### PR TITLE
Setup release process

### PR DESCRIPTION
**Pull request by smacker**
Fixes: https://github.com/src-d/code-annotation/issues/8

TODO (need infra):

* [x] Move from quay.io to docker hub
* [x] Test on minikube
* [x] Test helm chart on cluster with ingress

### CODE REVIEW

**Comment by bzz** [comment]
AFAIK current setup is copied from https://github.com/src-d/landing/

\cc @rporres for next steps

File: .drone.yml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    skip_tls_verify: true
+    chart: ./helm-charts/code-annotation
+    release: code-annotation
+    prefix: PROD
+    secrets: [ PROD_API_SERVER, PROD_KUBERNETES_TOKEN ]
+    values: ingress.globalStaticIpName=code-annotation-production,ingress.hostname=code-annotation.sourced.tech,image.tag=${DRONE_TAG}
+    tiller_ns: kube-system
+    wait: true
+    when:
+      event: [tag]
```
Comment (dpordomingo) [inline .drone.yml]: missing EOL

File: .drone.yml [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    when:
+      event: [pull_request]
+
+
+  # deployment to staging environment
+
+  build_stg:
+    image: golang:1.8-alpine3.6
+    environment:
+      - REACT_APP_SERVER_URL="//code-annotation.srcd.run"
```
Comment (dpordomingo) [inline .drone.yml]: expected url would be `//code-annotation-staging.srcd.run`
same pattern than for: `talks-staging.srcd.run`, `blog-staging.srcd.run`, and `landing-staging.srcd.run`
https://github.com/src-d/guide/blob/master/engineering/continuous-delivery.md

File: .drone.yml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      event: [push]
+
+  helm_deploy_stg:
+    image: quay.io/ipedrazas/drone-helm
+    skip_tls_verify: true
+    chart: ./helm-charts/code-annotation
+    release: code-annotation
+    prefix: STG
+    secrets: [ STG_API_SERVER, STG_KUBERNETES_TOKEN ]
+    values: ingress.globalStaticIpName=code-annotation-staging,ingress.hostname=code-annotation.srcd.run,image.tag=commit-${DRONE_COMMIT_SHA:0:7}
```
Comment (dpordomingo) [inline .drone.yml]: expected ingress.hostname would be expected url would be `code-annotation-staging.srcd.run`

**Review by dpordomingo — COMMENTED** [review]
I only performed a high-level review, and saw that the changes made are only:
- removing not needed services: landingApi and slackin,
- renaming landing by code-annotation
what makes sense for code-annotation necessities.

I only spotted some inconsistencies regarding hostnames for staging, so in general LGTM

File: .drone.yml (hunk 3 above)
Comment (smacker) [reply .drone.yml]: sure if I change it, I'll change it everywhere

File: .drone.yml [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    debug: true
+    when:
+      branch: [staging]
+      event: [push]
+
+  docker_stg:
+    group: docker
+    image: plugins/docker
+    registry: quay.io
+    repo: quay.io/srcd/code-annotation
```
Comment (bzz) [inline .drone.yml]: Change to DockerHub

File: .drone.yml [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    tiller_ns: kube-system
+    wait: true
+    when:
+      branch: [staging]
+      event: [push]
+
+
+  # deployment to production environment
+
+  build_prod:
```
Comment (bzz) [inline .drone.yml]: remove production

File: helm-charts/code-annotation/values.yaml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+service:
+  type: NodePort
+  codeAnnotation:
+    externalPort: 8080
+    internalPort: 8080
+    name: code-annotation
+ingress:
+  annotations:
+    kubernetes.io/ingress.class: gce
+    kubernetes.io/tls-acme: "true"
```
Comment (rporres) [inline helm-charts/code-annotation/values.yaml]: There was no way for you to know this, but we're moving away from kube-lego to handle Let's Encrypt certificates as it has given us some grief in the past and it's no longer under development. We're using kube-cert-manager. See https://github.com/src-d/docs/tree/master/helm-charts/docs

File: helm-charts/code-annotation/values.yaml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+# If provided, these will be used by the deployment
+# nodeSelector: XXXXXXX
+
+image:
+  # same tag for everyone as it is how landing repo Makefile works
+  tag: master
+  pullPolicy: IfNotPresent
+  registry: docker.io/srcd
+  codeAnnotation:
+    name: code-annotation
```
Comment (rporres) [inline helm-charts/code-annotation/values.yaml]: Are you planning to have multiple images for this in the future? If not, this is not necessary. It was done that way in landing as deployment is composed of multiple containers. The standard way of declaring this is using `image.repository` and `image.tag`, e.g. https://github.com/src-d/docs/blob/master/helm-charts/docs/values.yaml#L6-L7 and https://github.com/src-d/docs/blob/master/helm-charts/docs/templates/deployment.yaml#L20

**Review by rporres — CHANGES_REQUESTED** [review]
Looks good. A couple of details to be addressed.

**Comment by smacker** [comment]
Addressed. PR waits for published docker image and cluster with ingress.

**Comment by smacker** [comment]
@rporres I just updated PR. There are secrets are required now.

And you can take image here:
https://hub.docker.com/r/smacker/code-annotation/tags/

**Review by dpordomingo — APPROVED** [review]

**Comment by rporres** [comment]
We have a problem with those secrets due to https://github.com/ipedrazas/drone-helm/issues/52. The only workaround is for the chart to be able to accept a secret already deployed in the cluster.

If you want to keep your chart flexible enough (which I think it is good) I would modify the chart to:

* only create the secret if `.Values.secrets.github_client` and its friends are present (you can add a condition surrounding the whole secret definition in `templates/secrets.yaml`)
* define a variable containing the name of the secret at the beginning of `templates/deployment.yaml` based on a condition that would test if an existing secret is given or secret values are provided. See [this](https://github.com/src-d/landing/blob/master/helm-charts/landing/templates/ingress.yaml#L1) as an example on how to define variables in a chart

The above solution would only work if the existing secret structure is the same as the secret already present in the cluster, so that would also need be documented.

I'm sorry that you've hit that nasty drone-helm bug

**Comment by bzz** [comment]
It seems that requested changes were done to work it round, so from conversation IRL \w @smacker  we are just waiting for @rporres have some spare cycles to help us finishing testing it - the last item from PR todo in description.

**Comment by rporres** [comment]
Now that I'm back to work I will take a look into this

**Comment by rporres** [comment]
ingress metadata is missing the line
```
stable.k8s.psg.io/kcm.class: {{ .Values.ingress.kcmClass }}
```

**Comment by rporres** [comment]
https://code-annotation-staging.srcd.run/ and https://code-annotation-staging.srcd.run/ work with the ingress fix above. Since we cannot force redirections in GCE ingress, it would be nice that the application did it

**Comment by rporres** [comment]
I still have to configure drone, so please do not merge it

**Comment by smacker** [comment]
`kcm.class` added. I also rebased and squshed commits.

**Review by rporres — APPROVED** [review]
Drone set up. Feel free to merge it.

**Review by bzz — APPROVED** [review]
🎉 nice!
Looks great to me.

**Pull request merged**